### PR TITLE
fix(sweep): recover expected restack conflicts

### PR DESCRIPTION
**Pull request by howarewoo**
## Goal

Recover expected Graphite restack conflicts during a review sweep without weakening conflict handling or descendant-verification safety.

## Summary

- Inspect every paused rebase state, unmerged index stage, combined diff, and descendant commit before resolving a conflict.
- Reconcile both PRs' intent, stage only resolved conflict paths, and continue the interrupted Graphite operation until the stack-scoped restack completes.
- Treat descendant code as untrusted: execute project-code verification only in an isolated sandbox with repository secrets and credentials removed and outbound networking disabled; otherwise block before execution.
- Preserve actionable blocker behavior for ambiguous, unsafe, unverifiable, verification-failing, continuation-failing, or otherwise failed restacks.
- Cover the recovery, verification-isolation, and blocker boundaries with contract tests.

### Automated

- [ ] `bash skills/woostack-sweep/scripts/tests/test-restack-conflict-contract.sh` — 25 passed, 0 failed
- [ ] `bash skills/woostack-sweep/scripts/tests/run-tests.sh` — 25 + 4 + 6 passed, 0 failed
- [ ] Previously recorded: `bash skills/woostack-doctor/scripts/doctor.sh .` — 0 errors, 12 config/store warnings

### Manual

**Before merge**

- [ ] Read the restack workflow in sequence and confirm `gt submit --stack` occurs only after all required `gt continue` operations succeed, while unresolved failures preserve the paused worktree and rebase state.
- [ ] Confirm descendant project-code verification runs only with the required credential and network isolation, and otherwise blocks before execution.

Spec: .woostack/fixes/2026-07-14-sweep-auto-resolve-restack-conflicts.md

### CODE REVIEW

**Comment by howarewoo** [comment]
* **#523** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/523?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/howarewoo/woostack/523?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: skills/woostack-sweep/SKILL.md [hunk 1]
```diff
@@ -91,11 +91,26 @@ no-blocking verdict resolves in a single pass (step 2):
...
+   rebase onto the new tip. **Never `gt sync` or a repo-wide restack** ([worktree
+   contract](../woostack-init/references/worktrees.md) §4/§6: a repo-wide restack collides with
+   any parallel run in flight).
+   - If the restack pauses on a conflict, first enumerate the paused state and every unmerged
+     index stage with `git status --short`, `git ls-files -u`, and `git diff --cc`; read the
+     descendant change being replayed with `git rebase --show-current-patch`. Reconcile the
+     already-reviewed lower-PR fix with that descendant change's intent across every conflicted
+     path, including rename/delete and non-text conflicts. Never resolve by choosing an entire
+     `ours` or `theirs` side; either choice can silently discard one PR's behavior.
+   - Run the smallest existing focused verification that covers every affected behavior. If it
```
Comment (howarewoo) [inline skills/woostack-sweep/SKILL.md]: **Unreviewed descendant code executes during restack**

The new recovery loop runs a focused verification inside the paused descendant worktree before that descendant PR reaches its review turn. A contributor who controls the descendant branch can make the selected test or build command, or code imported by it, execute an exfiltration payload; merely engineering a restack conflict causes the autonomous sweep to run that payload with the maintainer agent's local credentials and network access.

Fix: Run conflict verification in a sandbox with repository secrets removed and outbound network disabled; if such isolation is unavailable, require interactive approval before executing descendant-controlled code.

<sub>— <strong>HIGH · BLOCKING</strong> · flagged by the <code>security</code> agent</sub>

File: skills/woostack-sweep/SKILL.md [hunk 2]
```diff
@@ -141,9 +156,12 @@ table + "Needs you".
 
 A **blocker** = the cap or no-progress guard reached with **blocking findings still present**, a
 `woostack-review` error/hang **or any PR left without a review receipt for its HEAD (the review did
-not run — never substitute a self/structural review)**, a restack/rebase conflict, or an
-`address-comments` step that would touch the never-auto-approve set (destructive / secret / auth /
-network / ambiguous). Safety is never relaxed for autonomy.
+not run — never substitute a self/structural review)**, a restack conflict whose intent is
+ambiguous or unsafe or cannot be verified safely, focused verification fails, `gt continue` fails,
+another `gt restack` failure occurs, or an `address-comments` step would touch the
```
Comment (howarewoo) [inline skills/woostack-sweep/SKILL.md]: **Nit: Failure blocker branches lack regression coverage**

The new blocker contract distinguishes a failed focused verification and a non-conflict `gt restack` failure from the adjacent ambiguous, unverifiable, and `gt continue` failure cases, but the added contract test asserts only those adjacent cases. Removing either of these two safety branches would therefore leave the suite green while allowing the sweep to proceed after a failed verification or an otherwise failed restack.

Fix: Add explicit negative-path assertions that the blocker section retains both `focused verification fails` and the separate other-`gt restack`-failure outcome, alongside the existing continuation-failure assertion.

<sub>— <strong>MEDIUM · NIT</strong> · flagged by the <code>tests</code> agent</sub>

**Review by howarewoo — COMMENTED** [review]
## AI Deep Code Review Summary

The full swarm found one blocking trust-boundary issue in autonomous restack verification and one test-coverage nit. The blocking path can execute descendant-controlled verification code before that PR reaches review.

---
**Status: CHANGES REQUESTED** — 1 blocking + 0 non-blocking finding(s) (1 HIGH, 0 MEDIUM, 0 LOW) + 1 nit(s). See inline comments.
*Audited by woostack-review · Host: OMP · Provider: OpenAI · Model: openai-codex/gpt-5.6-sol*



_Review event downgraded to COMMENT — GitHub blocks REQUEST_CHANGES on your own PR. Status line above carries the actual verdict._

File: skills/woostack-sweep/SKILL.md (hunk 1 above)
Comment (howarewoo) [reply skills/woostack-sweep/SKILL.md]: Fixed in `0367e326bee685596f843f3e941d404ddfe6bbe6`. Descendant project code now runs only in an isolated sandbox with repository secrets and credentials removed and outbound networking disabled; unavailable isolation blocks before verification executes.

File: skills/woostack-sweep/SKILL.md (hunk 2 above)
Comment (howarewoo) [reply skills/woostack-sweep/SKILL.md]: Fixed in `0367e326bee685596f843f3e941d404ddfe6bbe6`. The contract now asserts focused-verification failures and other `gt restack` failures remain blockers.

File: skills/woostack-sweep/SKILL.md [hunk 3]
```diff
@@ -91,11 +91,30 @@ no-blocking verdict resolves in a single pass (step 2):
...
+     descendant change being replayed with `git rebase --show-current-patch`. Reconcile the
+     already-reviewed lower-PR fix with that descendant change's intent across every conflicted
+     path, including rename/delete and non-text conflicts. Never resolve by choosing an entire
+     `ours` or `theirs` side; either choice can silently discard one PR's behavior.
+   - Treat descendant code as untrusted. Execute project code for focused verification only inside
+     an isolated sandbox with repository secrets and credentials removed and outbound networking
+     disabled. If the host cannot guarantee both controls, stop as a blocker without executing
+     verification. Otherwise run the smallest existing focused verification that covers every
+     affected behavior. If it passes, stage only the resolved conflict paths with
+     `git add -- <paths>`, then run `gt continue`. Repeat the inspect → reconcile → verify →
```
Comment (howarewoo) [inline skills/woostack-sweep/SKILL.md]: **Conflict paths can execute shell commands**

The conflicted path names come from an untrusted descendant branch, but the workflow tells the agent to substitute them into `git add -- <paths>`. A contributor can create a conflicting filename containing shell syntax such as `;` or `$()`; when the autonomous agent renders that filename into its shell command, the shell evaluates the syntax before Git sees the arguments. The `--` protects only Git option parsing, so this gives the contributor arbitrary command execution on the maintainer host during the restack, outside the verification sandbox.

Fix: Require NUL-delimited path discovery and staging without interpolating path text into a shell command, for example by piping `git diff --name-only --diff-filter=U -z` into `git add --pathspec-from-file=- --pathspec-file-nul`, and explicitly treat displayed filenames as untrusted data rather than executable command text.

<sub>— <strong>HIGH · BLOCKING</strong> · flagged by the <code>security</code> agent</sub>

**Review by howarewoo — COMMENTED** [review]
## AI Deep Code Review Summary

The full review found one blocking command-injection risk in the autonomous conflict-recovery instructions. Conflict pathnames must remain data and never be interpolated into shell commands.

---
**Status: CHANGES REQUESTED** — 1 blocking + 0 non-blocking finding(s) (1 HIGH, 0 MEDIUM, 0 LOW). See inline comments.
*Audited by woostack-review · Host: omp · Provider: openai · Model: openai-codex/gpt-5.6-sol*



_Review event downgraded to COMMENT — GitHub blocks REQUEST_CHANGES on your own PR. Status line above carries the actual verdict._

File: skills/woostack-sweep/SKILL.md (hunk 3 above)
Comment (howarewoo) [reply skills/woostack-sweep/SKILL.md]: Fixed in `f09e1430eed5a6b4daa65c36361ca3a74c03177f`. Conflict paths now flow NUL-delimited into `git --literal-pathspecs add --pathspec-from-file=- --pathspec-file-nul`; displayed filenames are explicitly treated as untrusted data and never interpolated into shell commands.

File: skills/woostack-sweep/scripts/tests/test-restack-conflict-contract.sh [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  "missing isolation blocks before descendant code executes"
+assert_contains "$restack_flat" 'git diff --name-only --diff-filter=U -z | git --literal-pathspecs add --pathspec-from-file=- --pathspec-file-nul' \
+  "restack recovery stages conflict paths without shell interpolation or pathspec magic"
+assert_contains "$restack_flat" '`gt continue`' \
+  "restack recovery resumes the halted Graphite command"
+assert_contains "$restack_flat" 'Repeat' \
+  "later descendant conflicts re-enter the recovery loop"
+assert_contains "$restack_flat" 'before `gt submit --stack`' \
+  "stack submission follows successful conflict continuation"
+assert_contains "$restack_flat" 'Never resolve by choosing an entire `ours` or `theirs` side' \
```
Comment (howarewoo) [inline skills/woostack-sweep/scripts/tests/test-restack-conflict-contract.sh:40]: **Nit: Edge conflict classes are not protected**

The recovery contract explicitly covers rename/delete and non-text conflicts, but this test only checks generic reconciliation and the prohibition on whole-side selection. Removing those edge-case guarantees from `SKILL.md` would leave all 25 assertions green, so the suite does not protect the conflict classes most likely to need special handling.

Fix: Add assertions over `restack_flat` that require the rename/delete and non-text conflict guarantees, so deleting either scope from the recovery contract fails this test.

<sub>— <strong>MEDIUM · NIT</strong> · flagged by the <code>tests</code> agent</sub>

**Review by howarewoo — COMMENTED** [review]
## AI Deep Code Review Summary

The full re-review found no blocking defects after the conflict-path fix. One non-blocking test-coverage nit remains for the documented rename/delete and non-text conflict guarantees.

---
**Status: APPROVED** — No blocking findings, 1 nit(s). See inline comments.
*Audited by woostack-review · Host: omp · Provider: openai · Model: openai-codex/gpt-5.6-sol*



_Review event downgraded to COMMENT — GitHub blocks APPROVE on your own PR. Status line above carries the actual verdict._

File: skills/woostack-sweep/scripts/tests/test-restack-conflict-contract.sh (hunk 4 above)
Comment (howarewoo) [reply skills/woostack-sweep/scripts/tests/test-restack-conflict-contract.sh:40]: Fixed in `7081d07c09a9b066237e54474efdf4015194af75`. The contract test now independently requires both `rename/delete` and `non-text conflicts`; the focused test reports 27 passed, 0 failed.
